### PR TITLE
feat(pr-review): persist skipped findings + learned review guidelines

### LINKED ISSUE #575: feat(pr-review): persist skipped findings + learned review guidelines
## Problem

During `/pr-review`, 3 parallel reviewers produce findings which are deduplicated and presented to the user for selection. The user picks which findings to post as GitHub PR comments (e.g., "post 1 and 3, skip 2").

Only posted comments are stored in the review database (`.pi/data/pr-reviews.db`). Skipped findings are discarded — no record exists that the user saw a finding and chose not to post it.

In the next review cycle on the same PR, the same reviewers may raise the identical finding again. The user is forced to re-evaluate and re-skip it every cycle.

Additionally, when a user repeatedly skips the same *type* of finding across PRs (e.g., "don't flag snake_case naming"), reviewers have no way to learn this preference and keep re-raising it.

### Part 1: Persist skipped findings (same-PR dedup)

- Add `status` + `skip_reason` columns to `pr_comments` table in `pr_review_store.py`
- Phase 4: when user skips findings, ask for reason, AI refines it (concise, technical)
- Phase 5b: store ALL findings — posted (`status='posted'`) and skipped (`status='skipped'` + refined reason)
- Phase 8: query DB for previously skipped findings, auto-skip matches via body similarity

### Part 2: Learned review guidelines (cross-PR, per-repo)

- AI classifies each skip reason as finding-specific or generalizable (no user interaction)
- Generalizable dismissals → appended to `.pi/data/review-guidelines.md` (gitignored, per-repo, per-user)
- All 3 code-reviewer agents check for this file before reviewing and respect its guidelines
- Works for ALL review flows (pr-review, review-handler, manual reviews, code review loop)

## Affected Files

- `myk_pi_tools/pr/pr_review_store.py` — schema + migration + queries
- `prompts/pr-review.md` — Phase 4/5b/8 changes
- `agents/code-reviewer-quality.md` — read review-guidelines.md
- `agents/code-reviewer-guidelines.md` — read review-guidelines.md
- `agents/code-reviewer-security.md` — read review-guidelines.md
- `AGENTS.md` — document new mechanism
- `tests/python/` — new tests

## Done

- [ ] Add `status` + `skip_reason` columns with migration
- [ ] Add `get_skipped_comments()` query
- [ ] Update Phase 4: derive skipped set → ask reason → refine → classify
- [ ] Update Phase 5b: store all findings (posted + skipped)
- [ ] Update Phase 8: query and auto-skip previously dismissed findings
- [ ] Write generalizable dismissals to `.pi/data/review-guidelines.md`
- [ ] Add review-guidelines check to all 3 reviewer agents
- [ ] Write tests
- [ ] Update AGENTS.md

### PR DESCRIPTION
**Pull request by myakove**
## Summary

Closes #575

When using `/pr-review`, skipped findings are now persisted in the database and used for same-PR dedup in subsequent cycles. Additionally, generalizable skip reasons are written to `.pi/data/review-guidelines.md` — a per-repo, gitignored file that all 3 reviewer agents read before reviewing, preventing the same class of finding from being raised again.

### Python (`myk_pi_tools/pr/`)

- **`pr_review_store.py`** — Added `status` + `skip_reason` columns to `pr_comments` table and `author` column to `pr_reviews` table, with schema migration for existing DBs. Added `get_skipped_comments()` query. Added `VALID_STATUSES` validation.
- **`commands.py`** — Added `myk-pi-tools pr get-skipped-comments {owner} {repo} {pr_number}` CLI subcommand.

### Prompts

- **`pr-review.md`** — Phase 4: derive skipped set → ask reason → AI refines → classify (finding-specific vs generalizable) → write to guidelines file. Phase 5b: store ALL findings (posted + skipped) with author. Phase 8: query and auto-skip previously dismissed findings via CLI.
- **`refine-review.md`** — Added `author` field to store metadata.

### Agents

- **All 3 code-reviewer agents** — Added "Learned Review Preferences" section: check `.pi/data/review-guidelines.md` before reviewing.

### Docs & Tests

- **`AGENTS.md`** — Documented PR Review Store changes and Learned Review Preferences.
- **`tests/python/test_pr_review_store.py`** — 9 tests covering store, query, migration, validation.

## Done

- [x] Add `status` + `skip_reason` columns with migration
- [x] Add `author` column with migration
- [x] Add `get_skipped_comments()` query + CLI subcommand
- [x] Add status validation
- [x] Update Phase 4: derive skipped set → ask reason → refine → classify
- [x] Update Phase 5b: store all findings (posted + skipped) with author
- [x] Update Phase 8: query and auto-skip previously dismissed findings
- [x] Write generalizable dismissals to `.pi/data/review-guidelines.md`
- [x] Add review-guidelines check to all 3 reviewer agents
- [x] Write tests (9 passing)
- [x] Update AGENTS.md

### CODE REVIEW

File: myk_pi_tools/pr/pr_review_store.py [hunk 1]
```diff
@@ -124,23 +164,34 @@ def store_pr_review(
...
         review_id = cursor.lastrowid
         if not review_id:
             raise RuntimeError("Failed to insert pr_reviews record")
 
+        posted_count = 0
+        skipped_count = 0
         for comment in comments:
+            status = comment.get("status", "posted")
+            if status not in VALID_STATUSES:
+                raise ValueError(f"Invalid status '{status}', must be one of {VALID_STATUSES}")
```
Comment (myakove) [inline myk_pi_tools/pr/pr_review_store.py]: [WARNING] **ValueError mid-transaction + uncaught in run_store()**

If a comment list has `[valid, valid, invalid_status]`, the first two are inserted before `ValueError` is raised. The `except sqlite3.Error` handler doesn't catch `ValueError`, so `conn.rollback()` is never called.

Also, `run_store()` only catches `RuntimeError` — the `ValueError` from status validation propagates as an unhandled exception with a raw traceback.

**Fix:** Validate all statuses before starting inserts, or catch `ValueError` in `run_store()` alongside `RuntimeError`.

```suggestion
    try:
        store_pr_review(owner, repo, pr_number, comments, head_sha=head_sha, author=author)
    except (RuntimeError, ValueError) as e:
```

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/reviews/fetch.py [hunk 2]
```diff
@@ -1414,6 +1414,12 @@ def run(
 
         print_stderr(f"Repository: {owner}/{repo}, PR: {pr_number}")
 
+        # Fetch PR author
+        pr_author: str | None = None
+        author_data = run_gh_api(f"/repos/{owner}/{repo}/pulls/{pr_number}")
+        if isinstance(author_data, dict):
+            pr_author = author_data.get("user", {}).get("login") if author_data.get("user") else None
```
Comment (myakove) [inline myk_pi_tools/reviews/fetch.py:1421]: [WARNING] **Extra API call for PR author with silent failure**

This adds a full `/repos/{owner}/{repo}/pulls/{pr_number}` API call solely for the author login. When it fails (rate limit, permissions), `pr_author` silently becomes `None` with no indication.

**Fix:** Add a warning log on failure:

```suggestion
        if isinstance(author_data, dict):
            pr_author = author_data.get("user", {}).get("login") if author_data.get("user") else None
        else:
            print_stderr(f"Warning: Could not fetch PR author (API returned {type(author_data).__name__})")
```

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/python/test_pr_review_store.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+from __future__ import annotations
+
+import sqlite3
+from pathlib import Path
+
+import pytest
+
+from myk_pi_tools.pr.pr_review_store import (
+    get_skipped_comments,
+    store_pr_review,
```
Comment (myakove) [inline tests/python/test_pr_review_store.py:16]: [WARNING] **Missing test for VALID_STATUSES ValueError path**

The `VALID_STATUSES` validation (`ValueError` on invalid status) at `pr_review_store.py:183-184` has no test coverage.

**Add:**
```python
def test_store_invalid_status_raises(db_path):
    with pytest.raises(ValueError, match="Invalid status"):
        store_pr_review("org", "repo", 1, [{"body": "x", "status": "invalid"}], head_sha="sha")
```

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/pr_review_store.py [hunk 4]
```diff
@@ -84,23 +98,49 @@ def _get_db_path() -> Path:
...
 
 
+def _run_migrations(conn: sqlite3.Connection) -> None:
+    """Apply schema migrations for existing databases."""
+    # Cache column info per table
+    table_columns: dict[str, set[str]] = {}
+
+    for table, col_name, sql in _MIGRATIONS:
+        if table not in table_columns:
+            cursor = conn.execute(f"PRAGMA table_info({table})")
```
Comment (myakove) [inline myk_pi_tools/pr/pr_review_store.py:109]: [SUGGESTION] **Add trust boundary comment on PRAGMA f-string**

The `f"PRAGMA table_info({table})"` is safe because `table` comes from the hardcoded `_MIGRATIONS` constant. PRAGMA doesn't support parameterized queries. Add a comment noting this for future maintainers.

```suggestion
            # table names come from hardcoded _MIGRATIONS — safe, PRAGMA can't be parameterized
            cursor = conn.execute(f"PRAGMA table_info({table})")
```

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Review by myakove — COMMENTED** [review]
## Code Review

Found **4** issue(s) in this PR:

### :bulb: Suggestions (4)

| File | Line | Issue |
|------|------|-------|
| `myk_pi_tools/pr/pr_review_store.py` | 184 | [WARNING] **ValueError mid-transaction + uncaught in run_store()** |
| `myk_pi_tools/reviews/fetch.py` | 1421 | [WARNING] **Extra API call for PR author with silent failure** |
| `tests/python/test_pr_review_store.py` | 12 | [WARNING] **Missing test for VALID_STATUSES ValueError path** |
| `myk_pi_tools/pr/pr_review_store.py` | 108 | [SUGGESTION] **Add trust boundary comment on PRAGMA f-string** |

---
*Review generated by pi*

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/pr_review_store.py (hunk 1 above)
Comment (myakove) [reply myk_pi_tools/pr/pr_review_store.py]: Fixed in commit f2750fb.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/reviews/fetch.py (hunk 2 above)
Comment (myakove) [reply myk_pi_tools/reviews/fetch.py:1421]: Fixed in commit f2750fb.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/python/test_pr_review_store.py (hunk 3 above)
Comment (myakove) [reply tests/python/test_pr_review_store.py:16]: Fixed in commit f2750fb.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/pr_review_store.py (hunk 4 above)
Comment (myakove) [reply myk_pi_tools/pr/pr_review_store.py:109]: Fixed in commit f2750fb.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/pr_review_store.py [hunk 5]
```diff
@@ -124,23 +164,34 @@ def store_pr_review(
...
                 """
                 INSERT INTO pr_comments (
-                    review_id, thread_id, comment_id, path, line, body, severity, posted_at
-                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+                    review_id, thread_id, comment_id, path, line, body, severity,
+                    posted_at, status, skip_reason
+                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                 """,
                 (
                     review_id,
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/pr/pr_review_store.py:205]: Fixed: status=skipped now requires non-empty skip_reason. Validation added before DB inserts.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/pr_review_store.py [hunk 6]
```diff
@@ -164,12 +220,53 @@ def store_pr_review(
...
+            ORDER BY c.id
+            """,
+            (owner, repo, pr_number),
+        )
+        return [dict(row) for row in cursor.fetchall()]
+    except sqlite3.Error as e:
+        log(f"Warning: Failed to query skipped comments: {e}")
+        return []
+    finally:
+        conn.close()
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/pr/pr_review_store.py:268]: Fixed: added conn.commit() after _run_migrations() in get_skipped_comments().

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: myk_pi_tools/pr/commands.py [hunk 7]
```diff
@@ -83,6 +83,26 @@ def pr_store_review(json_file: str) -> None:
...
+    """
+    import json
+    import sys
+
+    from myk_pi_tools.pr.pr_review_store import get_skipped_comments
+
+    results = get_skipped_comments(owner, repo, pr_number)
+    print(json.dumps(results, indent=2))
+    sys.exit(0)
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/pr/commands.py:108]: Fixed: get_skipped_comments() now raises RuntimeError on DB errors. CLI catches it and exits non-zero.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `myk_pi_tools/pr/commands.py:86` (qodo bug) — Skipped query masks failures
> Addressed: Fixed: get_skipped_comments() now raises RuntimeError on DB errors. CLI catches it and exits non-zero.
### `myk_pi_tools/pr/pr_review_store.py:181` (qodo requirement gap) — skipped allows null skip_reason
> Addressed: Fixed: status=skipped now requires non-empty skip_reason. Validation added before DB inserts.
### `myk_pi_tools/pr/pr_review_store.py:240` (qodo bug) — Uncommitted schema migration
> Addressed: Fixed: added conn.commit() after _run_migrations() in get_skipped_comments().
### `myk_pi_tools/pr/commands.py:86` (qodo finding) — pr_get_skipped missing test
> Addressed: Fixed: added test_cli_get_skipped using CliRunner.
### `myk_pi_tools/pr/pr_review_store.py:263` (qodo finding) — run_store() author untested
> Addressed: Fixed: added test_run_store_with_author.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/python/test_pr_review_store.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    """Passing status='skipped' without skip_reason raises ValueError."""
+    comments = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped"}]
+    with pytest.raises(ValueError, match="requires a non-empty skip_reason"):
+        store_pr_review("org", "repo", 1, comments, head_sha="sha1")
+
+    # Also test with empty / whitespace-only skip_reason
+    comments_ws = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped", "skip_reason": "  "}]
+    with pytest.raises(ValueError, match="requires a non-empty skip_reason"):
+        store_pr_review("org", "repo", 2, comments_ws, head_sha="sha2")
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/python/test_pr_review_store.py:279]: Fixed: split into test_store_skipped_without_reason_raises and test_store_skipped_whitespace_reason_raises — one behavior per test.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `tests/python/test_pr_review_store.py:265` (qodo finding) — test_store_skipped_without_reason_raises mixes cases
> Addressed: Fixed: split into test_store_skipped_without_reason_raises and test_store_skipped_whitespace_reason_raises — one behavior per test.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/python/test_pr_review_store.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    comments = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped"}]
+    with pytest.raises(ValueError, match="requires a non-empty skip_reason"):
+        store_pr_review("org", "repo", 1, comments, head_sha="sha1")
+
+
+def test_store_skipped_whitespace_reason_raises(db_path: Path) -> None:  # noqa: ARG001
+    """Passing status='skipped' with whitespace-only skip_reason raises ValueError."""
+    comments = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped", "skip_reason": "  "}]
+    with pytest.raises(ValueError, match="requires a non-empty skip_reason"):
+        store_pr_review("org", "repo", 2, comments, head_sha="sha2")
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/python/test_pr_review_store.py]: Fixed: removed match= from pytest.raises (test behavior not wording) and broke long dict literal into multi-line.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/python/test_pr_review_store.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+def test_store_skipped_without_reason_raises(db_path: Path) -> None:  # noqa: ARG001
+    """Passing status='skipped' without skip_reason raises ValueError."""
+    comments = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped"}]
+    with pytest.raises(ValueError, match="requires a non-empty skip_reason"):
+        store_pr_review("org", "repo", 1, comments, head_sha="sha1")
+
+
+def test_store_skipped_whitespace_reason_raises(db_path: Path) -> None:  # noqa: ARG001
+    """Passing status='skipped' with whitespace-only skip_reason raises ValueError."""
+    comments = [{"path": "a.py", "line": 1, "body": "needs reason", "status": "skipped", "skip_reason": "  "}]
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/python/test_pr_review_store.py]: Fixed: removed match= from pytest.raises (test behavior not wording) and broke long dict literal into multi-line.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `tests/python/test_pr_review_store.py:274` (qodo finding) — Brittle pytest.raises(..., match=...)
> Addressed: Fixed: removed match= from pytest.raises (test behavior not wording) and broke long dict literal into multi-line.
### `tests/python/test_pr_review_store.py:274` (qodo bug) — Ruff line-length violation
> Addressed: Fixed: removed match= from pytest.raises (test behavior not wording) and broke long dict literal into multi-line.

---
*Assisted-by: PI (claude-opus-4-6-1m)*
